### DISCUSSION TITLE
Loc as a foundational package

### DISCUSSION THREAD

**Discussion by mbarbin** [ocaml/dune#10948] [Ideas]
I am extracting here the part of this discussion https://github.com/ocaml/opam-repository/issues/26604 about releasing a loc package to opam, as it relates to dune:

I am using a type to represent a range of lexing positions from a file, which I called Loc.t. It's [here](https://github.com/mbarbin/loc).

I use it as dependency to decorate AST nodes built by parsers, enabling located error messages during file processing (compilers, interpreters, linters, refactor tools, etc.).

I am considering submitting it to the opam-repository, but I have some reservations.

It is very similar to Dune's `Stdune.Loc` module (and actually uses `stdune` under the hood).

Perhaps it would make more sense to work on extracting the Dune one as a foundational package (such as pp, dyn, and ordering), rather than maintaining this one separately.

Would you have interest in entertaining such extraction, and make what's currently `Stdune.Loc` a standalone & reusable package?

**Comment by rgrinberg** [comment]
Yeah, why not I suppose. If you think our Loc type is good enough for your use cases, we can extract it into a separate package. I think you'll need to lead the effort on that though.

**Comment by Leonidas-from-XIV** [comment]
I guess this is solved, given `ocamlc-loc` exists as a separate package in Dune?

**Reply by rgrinberg** [reply]
ocamlc-loc is just for parsing locations as outputted by the compiler.

**Reply by mbarbin** [reply]
I added a list of related libraries/modules that I vaguely know of as a comment. Thank you for joining the conversation!

**Comment by mbarbin** [comment]
In the loc project README I had started to list similar libraries and/or modules from existing projects. I've come across a couple more since. I'll start a list here, please feel free to extend I may easily have missed others.


- Dune's [Loc.t](https://github.com/ocaml/dune/blob/dedfb76837bad6ea6de97de07e61f2ac10442127/otherlibs/stdune/src/loc.mli#L1)
- Parsexp's [ranges](https://github.com/janestreet/parsexp/blob/14af9ab942251783de6abb20e0d0e0eec6080062/src/positions.mli#L59)
- Ppxlib's [Location.t](https://github.com/ocaml-ppx/ppxlib/blob/456d1a99d354f8b3f34f01d1a7b61dc43ef678b6/src/location.ml#L4)
- Compilerlibs's [loc](https://github.com/ocaml/ocaml/blob/81003f23918729fa730c01ccffec25939cd730ff/utils/warnings.mli#L23)
- Jsont's [Texloc](https://github.com/dbuenzli/jsont/blob/b5575e6730c9d358a9bd02a3136440c5f7a1691e/src/jsont.mli#L31)
- Publicodes's [Pos](https://github.com/publicodes/publicodes/blob/c1888040ac431ae5a381b4d905d5dc6f477eb9c3/packages/compiler/lib/utils/pos.mli#L24)
- [Pp_loc](https://github.com/Armael/pp_loc)

Also, recently I saw that the [Zanuda](https://github.com/Kakadu/zanuda) OCaml Linter was able to output located error messages from OCaml source code including quoted text, like dune. I haven't looked what lib powers this under the hood, maybe some other, or could be compiler's lib?
